### ISSUE TITLE
Bug: [no-unused-vars] Single unused variable not flagged among others

### ISSUE THREAD

**Issue by fmaiabatista** [typescript-eslint/typescript-eslint#8418]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.3.3&fileType=.tsx&code=PQKgBApgzgNglgOwC4FoAmcoEMBGMJgACSAngA7QDGATnGatPMsAgPYoCuC2AZhCpSwwYAGiKkKUGnQaxESFuy69%2BAWwiqcEaiiyVK0KGLacEaCDzAhgAKDiqyrakjABvMOZ6IIAYVYIvAHMwAF8wHmpWVTAAcgA3OCQIGLsHJxdqCD0XCKjYwgSkgCsoYDIYDkDEFEzslJtgYDAACyQkMigALkbCiBKAOnM44Ep-INsIAA9HZw8LLA4YF09vPwC4QIAKVxswMHLKxC6wAG1ayiRNgEoAXRFdsChtOO1Otwe9skjJkjedvYBsR4rFYMT%2BH0BYCQWGogQgSDeMWBoPukIBlGaWAQcIA8rQqgg3khqBwIKi0WBRutAhxMm9Nl9WD8xKx6HB-FArmAALwAPneFIBjJ%2B-X8mxi2ki1BiYk2krEmQAjgroFy%2BQLBeiOax8P0YKwtjEAArfEiQahSmXm6hXCFokK2zX7U2ihDi4UkABKEEVVoZpu9yrASpVnJ5-P%2BTsp2t1%2BsNAGUIGZEMElaSoC4kKwoc0CNDYfDOlalf11Ehmqw0CrFf1aTBHZqHXbIR7Xe6A9A-R7vUZgz7Q2qI820VSoDqIHqDeLvQY4C80H2oI5uARctFy3mYXCEVbu9B%2BhmsEgOFA-OZq7XqPXhwCm4KQuTAQ%2BIc%2B9s%2Bm0A&eslintrc=N4KABGBEBOD2sBdIC4wOgVwKYBpxSwDsA3FMUCCSQ2AEyzPW3wF88qsAPBI2gZzIBtfBz4AbAJaEEyaFgDGsALZLeWWpHaVIABzEYA5lOQABBAE8dWPvOgSdCALTXJ0gPRzFKtbUcWrjvIAFgoA1uqQ%2BAC6WpASBjRyAAoAhgg80IQCqMKUULQSfEhaVAB0LlLo8qXyAFYC0bE6KdB8WNBkkGaW1rb2ThXuza3tmvi6LW3QAPIOErBZZBTaCkopAGrtfPOEnWJp1sUiUHywGNDyWAAqPZ1KdBhiDCVQOnC1Ckg5kKVuCDYLABm8VKNHopXqC0gMWOkH%2BikIwIMACV4AgACISDqoSAAfVxBUyKVUkQgLBALCAA&tsconfig=&tokens=false

### Repro Code

```TypeScript
/* eslint-disable @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access, no-undef */
import { defineConfig } from 'vite'
import react from '@vitejs/plugin-react'

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [react()],
  server: {
    proxy: {
      'foo': {
        target: 'foo',
        changeOrigin: true,
        configure: (proxy, options) => {
          proxy.on('error', (err, req, res) => {
            console.log('Proxy error', err)
          })
          proxy.on('proxyReq', (proxyReq, req, res) => {
            console.log('Sending request to the target:', req.method, req.url)
          })
          proxy.on('proxyRes', (proxyRes, req, res) => {
            console.log('Received response from the target:', proxyRes.statusCode, req.url)
          })
        },
      },
    },
  },
})
```

### ESLint Config

```javascript
{
  "root": true,
  "env": {
    "node": true
  },
  "extends": [
    "eslint:recommended",
    "plugin:@typescript-eslint/recommended-type-checked"
  ],
  "ignorePatterns": [
    "dist",
    ".eslintrc.cjs"
  ],
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": "latest",
    "sourceType": "module",
    "project": ["./tsconfig.node.json"],
    "tsconfigRootDir": "__dirname"
  }
}
```

### Expected Result

I expected that `proxyReq` on line 17 would trigger the rule "no-unused-vars".

### Actual Result

Other variables trigger the error (and have squiggly lines underneath), but this one specifically doesn't.

My local env:
![Xnip2024-02-09_15-16-59](https://github.com/typescript-eslint/typescript-eslint/assets/1574224/343eb895-754b-4089-b688-994ca224c8f1)

Playground:
![image](https://github.com/typescript-eslint/typescript-eslint/assets/1574224/eee40935-d0a0-43bc-af29-26cbad921b56)

**Comment by fmaiabatista** [comment]
💡 Noticed that even if the ESLint config is empty in the playground, the behavior is still observable...

**Comment by auvred** [comment]
https://eslint.org/docs/latest/rules/no-unused-vars#args-all

> `{ "args": "all" }`

> `all` - all named arguments must be used.



[playground](https://typescript-eslint.io/play/#ts=5.3.3&fileType=.ts&code=KYDwDg9gTgLgBAMwK4DsDGMCWEWJQClSQGdgATALjlQGsUIB3FAGmtMupTsZQEo4A3gCg4cNDmIQANsAB0UiAHNC7XkIC%2BQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQHYHsBaWXWZRAE0IDcBDaVDSfAMxcgBpxtIABAFwCeABxQBjaAEth-QiniTc-APQFipclToN0YANqRE0aPmidM3bFHoBzRtfjxIlsAF8Aut1chXQA&tsconfig=&tokens=false)

**Comment by bradzacher** [comment]
The default for the `args` option is `after-used`.
See the links @auvred posted for more info.

**Comment by fmaiabatista** [comment]
Thanks for the replies, this one tripped me up, sorry for missing that...

One thing to note is that in the final config that worked for me, I didn't have to turn off eslint's own `no-unused-vars` rule as done in the playground above - it was fine to simply have:

```
rules: {
  // ...
  '@typescript-eslint/no-unused-vars': ['error', { args: 'all', argsIgnorePattern: '^_' }]
}
```
